internal/dot/graph.go

```diff
@@ -20,53 +20,166 @@
 
 package dot
 
-import "fmt"
+import (
+	"fmt"
+	"reflect"
+)
 
 // Ctor encodes a constructor provided to the container for the DOT graph.
 type Ctor struct {
-	Name    string
-	Package string
-	File    string
-	Line    int
-	Params  []*Node
-	Results []*Node
+	Name       string
+	Package    string
+	File       string
+	Line       int
+	Params     []*Param
+	GroupParam []*Group
+	Results    []*Result
+}
+
+// Param is a parameter node in the graph.
+type Param struct {
+	*Node
+	Optional bool
+}
+
+// Result is a result node in the graph.
+type Result struct {
+	*Node
+	GroupIndex int
+}
+
+// Group is a group node in the graph.
+type Group struct {
+	Type    reflect.Type
+	Group   string
+	Results []*Result
 }
 
 // Graph is the DOT-format graph in a Container.
 type Graph struct {
-	Ctors []*Ctor
+	Ctors  []*Ctor
+	Groups map[key]*Group
 }
 
-// Node is a single node in a graph.
+// Node is a single node in a graph and is embedded into Params and Results.
 type Node struct {
-	Type     string
-	Name     string
-	Optional bool
-	Group    string
+	Type  reflect.Type
+	Name  string
+	Group string
+}
+
+type key struct {
```

Comment:
seeing as we only use `key` for the map of groups now, we might want to rename this `groupKey`.

Reply:
sure, `groupKey` definitely is more descriptive.